File: backtracking/combinations.cpp
class Solution {
public:
    vector<vector<int>> ans; 
    vector<int> path;
    vector<vector<int>> combine(int n, int k) {
        backtrack(1, k, n);
        return ans;
    }

    void backtrack(int start, int k, int n) {
        if (path.size() == k) {
                ans.push_back(path);
                return;
            }
        for (int i = start; i <= n; i++) {
            path.push_back(i);
            backtrack(i + 1, k, n);
            path.pop_back();
        }
    }
};

File: backtracking/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans, comb = [], []

        def backtrack(start):
            if len(comb) == k:
                # not ans.append(comb), the value will change after poping
                ans.append(comb[:])
                return
            
            for i in range(start, n+1):
                comb.append(i)
                backtrack(i+1)
                comb.pop()
        
        backtrack(1)
        return ans